ksy_style_guide.adoc

```diff
@@ -249,28 +248,36 @@ guesswork.
 
 * For simple non-repeated fields, use a simple singular form —
   i.e. `width`, `header`, `transaction_id`, `file`.
-* For repeated files (i.e. with `repeat: something`), use plular form
+* For a sequence (i.e. with `repeat: something`), use plular form
   — i.e. `files`, `transactions`.
 * Don't be overly verbose: use commonly understood abbreviations
   liberally, if it will improve readablity — i.e. `src_mac` or
   `src_mac_addr` instead of `source_media_access_control_address`
+* Repeated fileds which cannot be packed into a sequence should
+  have `id`s containing a number in the end. Numbers may have a
+  visually-obvious structure, like "the first digit is major, the second
+  one is minor".
 * For fields that are designed to be used to detect file type (AKA
-  "magic values"), use `magic` name, or, if there are several of them,
-  `magic1`, `magic2`, etc.
+  "magic values"), use `signature` or `magic` name, or, if there are
+  several of them, like `signature` or `magic0`, `magic1`, etc.
 * For reserved fields which are *known* to be unused, use `reserved`
-  name (or `reserved1`, `reserved2`, etc, if there are many of them)
+  name (or `reserved0`, `reserved1`, etc, if there are many of them)
 * For fields that designate *number / count* of something (in
-  particular, number of repetitions of some other structure), use
-  `num_` prefix and plular form — i.e. `num_questions`, `num_blocks`,
-  `num_nodes`
+  particular, number of repetitions of some other structure), use either
+  `_count` suffix or `count_of_` (what sounds better in your opinion)
+  prefix and a plular form — i.e. `count_of_questions`, `blocks_count`,
+  `nodes_count`
 * For fields that designate *offset* to some particular data structure,
-  use `ofs_` prefix and name of that data structure (as it would appear
-  in the file) — i.e. `ofs_block`, `ofs_queries`, `ofs_path`
+  use `_offset` suffix and name of that data structure (as it would
+  appear in the file) — i.e. `block_offset`, `queries_offset`, `path_offset`
 * For fields that designate *size* of some particular data structure
-  (in bytes or some other fixed units), use `len_` prefix and name of
-  that data structure — i.e. `len_block` (length of a single `block`
-  entry), `len_blocks` (total length of whole `blocks` array, made of
+  (in bytes or some other fixed units), use `_size` suffix and name of
+  that data structure — i.e. `block_size` (length of a single `block`
+  entry), `blocks_size` (total length of whole `blocks` array, made of
```

Comment:
So, it boils down to:

1. Prefix vs suffix
2. Long of short spelling

In my opinion, "prefix" is better and definitely more widespread, i.e. as in majority of languages that use prefixes to differentiate parameters vs instance members, or Apps variety of Hungarian notation. Prefix does not try to sound like "proper English", and it avoids lots of ambiguilities and spelling problems (like that "count" stuff I've demonstrated above). For example, there are tricky nouns in English (like "water" or "money" or "news" or "advice" or "hair"), which make proper English phrasing difficult.

Last, but not least, I definitely prefer shorter form of spelling. Given that we have to make "one size fits all or most" identifiers, we're not going to have full-length Java-style identifiers like "NormalBucketDistributionProgressionOfLongIntegerConstantsForHadoopNodeFactory" — (1) there are languages that have certain ID length limits, (2) a vast majority people I've interviewed do not like such long identifiers very much, (3) it's actually hard to work with them without a IDE with expression autocompletion and stuff — and we don't have one (and I'm not very keen about having IDE as the *only* way to work with the language).

So, all and all, clear 3 character prefix looks like a good trade-off for me — it's non-ambiguous, it makes it very easy to detect whatever this attribute is about by always looking at the beginning, and it fits concise C-like style well.

Reply:
>In my opinion, "prefix" is better and definitely more widespread, i.e. as in majority of languages that use prefixes to differentiate parameters vs instance members, or Apps variety of Hungarian notation.

It doesn't mean that we should do this too.

The good part in postfix is that since we read left-to-right all the properties of the same member look similar to each other. Especially when aligned. But with prefix they look too differrent. Also take in mid3 the argument with point.

>(1) there are languages that have certain ID length limits,

Valid.

>(2) a vast majority people I've interviewed do not like such long identifiers very much

Neither do I. But In the case of short enough identifiers, which is the majority of them, I prefer to have them readable. It os only a guide, not a strict requirement.

>(3) it's actually hard to work with them without a IDE with expression autocompletion and stuff

Valid. BTW some text editors, like Notepad++, have autocompletion which autocompletes any tokens.